### ISSUE TITLE
Missing Row Separators in Templating Export

### ISSUE THREAD

**Issue by Jmuccigr** [OpenRefine/OpenRefine#3955]
### To Reproduce
1. First, select the "Templating..." option under Export menu.
2. Use the default values for the json output.
3. Finally, observe that a number of "rows" (really records) at the start, are missing the "Row separator".

### Current Results
The first two exported items are missing a row separator:

```
{
  "rows" : [
    {
      "One" : "a",
      "Two" : "1",
      "Three" : "2"
    }    {
      "One" : null,
      "Two" : "3",
      "Three" : "4"
    },
    {
      "One" : "c",
      "Two" : "5",
      "Three" : "6"
    }
  ]
}
```

Also this data set also gets displayed with 2 (not 3) rows because the second row, with no value in the first column, doesn't get numbered.

### Expected Behavior

My data set should have three rows and the export should have all the row separators.
### Screenshots

![Screen Shot 2021-06-02 at 15 36 00](https://user-images.githubusercontent.com/615115/120541747-455c2e80-c3b8-11eb-9569-05fe2c51406a.png)

### Versions
 - Operating System: MacOS 10.15.7
 - Browser Version: Safari 14.1.1, FireFox 88.01
 - JRE or JDK Version: build 1.8.0_111-b14
 - OpenRefine: 3.4.1

### Datasets
A small reproducible example:

```
"One","Two","Three"
"a",1,2
"",3,4
"c",5,6
```

**Comment by thadguidry** [comment]
Hi @Jmuccigr - the Templating exporter is a Row's based exporter, not Records based (until a future enhancement changes some of that, which is somewhere on our backlog for the last 10 years or so).  We do give hints about this in the dialog itself such as:

![image](https://user-images.githubusercontent.com/986438/120543702-3a060480-c3b2-11eb-8937-dba5a8188e84.png)

Where would could probably relabel the drop down menu on Export to say something like "Row-based Templating..." but it might be better overall to have OpenRefine tell you something like "Hey, your in Records mode, and this Templating exporter is designed for Row's based data, please switch modes before trying the Row-based Templating exporter".

Anyways, to get around this issue for now, you can go back and switch to Rows mode instead of Records and then try again and it should work fine.
Let us know your thoughts also on how you envision perhaps a Records based Templating exporter to work or look like which can be completely different than the current Rows based Templating exporter.  That will help us for the future when we get around to doing a redesign and better things for Records as described in #2825

**Comment by thadguidry** [comment]
Ah, I just double-checked something... looks like if you remove the `,` character in the Row Separator input box and then put it back again, the preview updates and shows the row separator.  But again as I stated, this doesn't quite represent a good tree view in JSON of the Records hierarchy itself, but only pure Rows. So let us know your thoughts on a Records based Templating exporter and your needs, even if it's a doodle on the back of a napkin.

**Comment by Jmuccigr** [comment]
Ah, I didn't even see the row vs record mode. Thanks. (I see now that a missing first-column value will bump it into records mode.) That solves my problem.

Regarding terminology, I think what's confusing is the the "Row separator" appears between the converted **input** rows, not between the **output** rows, though I guess if I had been aware of the row/record dichotomy, it might have made more sense to me. Maybe something more generic like "Item separator" would avoid the issue all together.

Weirdly, even when you're in record mode, the "Row Template" gets applied to each actual row of the spreadsheet. OR just doesn't put in the separator, which seems a bit odd to me. Something like the alert you mention would be useful. I'd expect at least that, if not some kind of error.

**Comment by wetneb** [comment]
Yeah I cannot really think of a reason why rows inside of a record should not be separated like others. I think a sensible solution to this issue would be to simply add row separators there as well.

If we want to think bigger, we could also try to think how we could make it possible to genuinely exploit the records structure there? It is really not clear to me what it should look like ideally, but it is pretty clear to me that the current behavior is indeed quite confusing.

**Comment by thadguidry** [comment]
@wetneb I think looking to the database world for records processing is wise.  Where databases have gotten quite well over the last 10 years supporting records .  I have to databases for advanced records processing since they provide the support that currently OpenRefine does not well. (Tony and I had lots of ideas one afternoon, but requires lots of work of course)  Anyways, PostgreSQL is quite exceptional here.  I've often wanted the power of jsonpath in general to be available in OpenRefine as expressed before in #2186.  With some handwaving here for simplicity sake...If OpenRefine Records handing could equal that of PostgreSQL's jsonb (B-tree operations) and records handling then that should suffice. (for example, all of Owen's asks before are supported in something like PostgresSQL's wonderful json handling functions and operators (including recordsets, Strict and Lax modes, etc.)

Scroll down through the docs especially to see
`jsonb_to_record`
`jsonb_to_recordset`
`jsonb_object_keys`
etc. :
https://www.postgresql.org/docs/current/functions-json.html

I dreamed of one idea a few weeks ago where I wished that the new-architecture would magically have a button that said (Advanced Mode) and pulled down PostgreSQL and fired it up in the background as the datagrid runner to support better Records mode, specifically where PostgreSQL's [Composite Types (structure of a row or record)](https://www.postgresql.org/docs/current/rowtypes.html) are available and its JSON functions and operators are available to work with them.

**Comment by Jmuccigr** [comment]
Being a relative newcomer to OR, I would love it if it popped up an alert that said "Hey, your dataset is missing some values in the first column, so we're putting you in Records mode. If you want us to treat this dataset by the row, click here." (vel sim.).

I suspect that most people wouldn't create datasets with empty first column values, but you never know!
